packages/matter.js/src/behavior/definitions/level-control/LevelControlServer.ts

```diff
@@ -4,14 +4,517 @@
+    }
+
+    async #stepIntervalTick() {
+        if (!this.internal.currentTransitionData) {
+            this.internal.transitionIntervalTimer?.stop();
+            return;
+        }
+        const { changeRate, withOnOff, targetLevel } = this.internal.currentTransitionData;
+        const newLevel = (this.state.currentLevel ?? this.internal.minLevel) + changeRate;
+        if (newLevel <= this.internal.minLevel) {
+            logger.debug(`Stopping transition interval at minLevel: ${this.internal.minLevel}.`);
+            await this.setLevel(this.internal.minLevel, withOnOff);
+            this.internal.transitionIntervalTimer?.stop();
+            this.setRemainingTime(0);
+        } else if (newLevel >= this.internal.maxLevel) {
+            logger.debug(`Stopping transition interval at maxLevel: ${this.internal.maxLevel}.`);
+            await this.setLevel(this.internal.maxLevel, withOnOff);
+            this.internal.transitionIntervalTimer?.stop();
+            this.setRemainingTime(0);
+        } else {
+            // Check if we reached the targetLevel if there is one
+            if (targetLevel !== undefined) {
+                if (changeRate > 0 && newLevel >= targetLevel) {
+                    logger.debug(`Stopping transition interval at targetLevel: ${targetLevel}.`);
+                    await this.setLevel(targetLevel, withOnOff);
+                    this.internal.transitionIntervalTimer?.stop();
+                    this.setRemainingTime(0);
+                    return;
+                } else if (changeRate < 0 && newLevel <= targetLevel) {
+                    logger.debug(`Stopping transition interval at targetLevel: ${targetLevel}.`);
+                    await this.setLevel(targetLevel, withOnOff);
+                    this.internal.transitionIntervalTimer?.stop();
+                    this.setRemainingTime(0);
+                    return;
+                }
+            }
+            logger.debug(`Setting new level in transition interval: ${newLevel}.`);
+            await this.setLevel(newLevel, withOnOff);
+
+            // There is no definition on how often the remaining time should be updated, so we update it with every step
+            if (this.internal.currentTransitionData.changeRate > 0) {
+                this.setRemainingTime(Math.floor(Math.ceil((this.internal.maxLevel - newLevel) / changeRate) * 10));
+            } else {
+                this.setRemainingTime(Math.floor(Math.ceil((newLevel - this.internal.minLevel) / changeRate) * 10));
+            }
+        }
+    }
+
+    override async [Symbol.asyncDispose]() {
+        if (this.internal.transitionIntervalTimer?.isRunning) {
+            this.internal.transitionIntervalTimer.stop();
+        }
+        await super[Symbol.asyncDispose]?.();
+    }
+}
+
+export namespace LevelControlServerLogic {
+    export class Internal {
+        /** Minimum Level value which always is defined. */
+        minLevel = 0;
```

Comment:
Is the duplicate field just because state.minLevel could be nullish?  Could create this.minLevel getter that returns state.minLevel ?? 0.  Just to avoid the confusion of having it defined twice

Reply:
Yes in fact it is because of the nullish. So I need this value internally because the attribute itself can be null. I thought. Getter on "class" level will not work. I will try it. Good idea.

Reply:
Great, seems to work